[wallet] Remove Wallet dependencies from init.cpp

jnewbery opened the pull request:
This continues the work of #7965. This PR, along with several others, would remove the remaining dependencies from libbitcoin_server.a on libbitcoin_wallet.a.

To create the interface, I've just translated all the old init.cpp wallet function calls into an interface class. I've not done any thinking about whether it makes sense to change that interface by combining/splitting those calls. This is a purely internal interface, so there's no problem in changing it later.

jnewbery committed:
[wallet] Create wallet init interface.

jnewbery committed:
[wallet] Move wallet init functions into WalletInit class.

promag commented:
Will take a look.

jnewbery commented:
Thanks @promag - this is very rough-and-ready at the moment. It builds and passes tests, but is not ready for code review. I'd be very happy to get feedback about general concept though.

jnewbery commented:
I see that @theuni has opened #10756 to replace CNodeSignals with an interface class. Would that be a better approach here?

ryanofsky commented:
> I see that @theuni has opened #10756 to replace CNodeSignals with an interface class. Would that be a better approach here?

Yeah I was just about to make this suggestion. You could delete the `WalletInitSignals` struct and replace the `WalletInitSignals` global variable with a `std::unique_ptr<WalletInitInterface>` global. This would get rid of all the boost signals code and make the control flow simpler.

Also you should prefix global variables with `g_`. And I think this PR would be less error prone and simpler to review if you just left the init code in wallet.cpp for now and moved the relevant functions to walletinit.cpp in followup MOVEONLY PR.

jnewbery commented:
> I think this PR would be less error prone and simpler to review if you just left the init code in wallet.cpp for now and moved the relevant functions to walletinit.cpp in followup MOVEONLY PR.

Thanks @ryanofsky, you've convinced me - I'll split this PR into two. But I think I'll do it the other way round - first move the wallet initialization/destruction functions into their own `walletinit` translation unit, and then introduce the interface class in a second PR. The reason is that #10740 is built on walletinit.cpp so this PR and #10740 will no longer be interdependent.

jnewbery commented:
Rebased on master. I haven't swapped out the boost signals for an interface class yet. Still looking for high-level feedback on this approach.

jonasschnelli commented:
Concept ACK on the general direction, though unsure about the implementation (haven't checked the code in detail).

I was following a similar approach in #6008 (relatively old closed PR) (check commit: https://github.com/bitcoin/bitcoin/pull/6008/commits/fa6a5b28fcaa612c6512c6028184df6b346fdad4).

Maybe, init/the-core only knows about modules. Some signal/hooks would allow initialising, shutdown of such modules. The wallet could be one, the miner another one, etc.
It would also allow to run with multiple wallet implementations (my idea was once to replace BDB with an append only file-format and removing all legacy code,... put this in a second wallet implementation and label it experimental).

If possible, I would prefer a more generic approach (a module interface rather then a wallet interface).

TheBlueMatt commented:
Concept ACK, generally. Agree that I'd really prefer no boost::signals (especially in the header!).

MarcoFalke commented:
Needs rebase

PierreRochard commented:
Concept ACK

jnewbery commented:
Rebased. Feedback not yet addressed.

jnewbery committed:
[wallet] Use global g_wallet_init_interface to init/destroy the wallet.

This commit creates a global g_wallet_init_interface, which is created
in bitcoind and bitcoin-qt. g_wallet_init_interface is used to init
and destroy the wallet.

This removes the dependency from init.cpp on the wallet library.

jnewbery commented:
I've reimplemented this as an interface class. Should be ready for review.

ryanofsky commented on src/wallet/init.cpp:
```diff
@@ -290,25 +290,28 @@ bool OpenWallets()
 
-void StopWallets() {
+void WalletInit::StopWallets() {
+    if (gArgs.GetBoolArg("-disablewallet", DEFAULT_DISABLE_WALLET)) return;
```
In commit "[wallet] Move wallet init functions into WalletInit class."

Is adding this line here intentional? It doesn't seem to fit in with other changes in this commit, and seems unnecessary. Should add a comment describing rationale if it is useful to keep.

jnewbery replied:
Yes, removed. This is a vestige of a previous branch.

ryanofsky commented on src/walletinitinterface.h:
```diff
@@ -0,0 +1,33 @@
+class CScheduler;
+class CRPCTable;
+
+class WalletInitInterface {
```
In commit "[wallet] Create wallet init interface."

Just FYI, this PR is very similar to 6626e2fe114ce2aa861be1eeda84f0c9faa60527 from #10973 and the `WalletInitInterface` added here is basically identical to the `Chain::Client` interface added there:

https://github.com/bitcoin/bitcoin/blob/6626e2fe114ce2aa861be1eeda84f0c9faa60527/src/interface/chain.h#L18-L40

Probably would replace this interface with that more generic one if this is merged first.

jnewbery replied:
Thanks for the heads up!

ryanofsky commented on src/walletinitinterface.h:
```diff
@@ -5,6 +5,8 @@
 #ifndef WALLETINITINTERFACE_H
 #define WALLETINITINTERFACE_H
 
+#include <string>
```
In commit "[wallet] Use global g_wallet_init_interface to init/destroy the wallet."

This change would make a little more sense in the previous commit adding the file.

jnewbery replied:
Agree. Fixed

ryanofsky reviewed (commented):
Looks good, conditional utACK 89d1d6ca293f602b88bf47fc2c3da37ddf6d4650 if you address the WalletInit::StopWallets comment below and also update the PR description to no longer refer to boost signals.

MarcoFalke commented on src/walletinitinterface.h:
```diff
@@ -0,0 +1,35 @@
+class WalletInitInterface {
+public:
+    /** Get wallet help string */
+    virtual std::string GetWalletHelpString(bool showDebug) {return (std::string)"";}
```
style-nit: Instead of casting a literal, you could directly return an empty `std::string{};`

jnewbery replied:
better. Thanks!

jnewbery commented:
Updated PR description to no longer reference boost signals.

This is ready for review. It's a fairly small changeset, and removes one of the last server->wallet dependencies.

jimpo commented on src/init.cpp:
```diff
@@ -415,9 +415,9 @@ std::string HelpMessage(HelpMessageMode mode)
-    strUsage += GetWalletHelpString(showDebug);
-#endif
+    if (g_wallet_init_interface) {
+        g_wallet_init_interface->GetWalletHelpString(showDebug);
```
This should be `strUsage += ...` I believe.

jnewbery replied:
You're right. Fixed

sipa commented:
Concept ACK

jimpo commented on src/walletinitinterface.h:
```diff
@@ -0,0 +1,35 @@
+class CScheduler;
+class CRPCTable;
+
+class WalletInitInterface {
```
Every method name in this interface has "Wallet" or "Wallets" in it, which is rather redundant given that this is a **Wallet**InitInterface. I'd say drop them all, so "GetWalletHelpString" => "GetHelpString", "WalletParameterInteraction" => "CheckParameterInteraction", "RegisterWalletRPC" => "RegisterRPC".

That alone would go a long way towards making this a generalized init module interface.

jnewbery replied:
I've updated the function names, since as you point out, "Wallet" is redundant.

See my comment below about a generalized module init interface.

jimpo reviewed (commented):
Definitely an improvement, though I agree with @jonasschnelli that I'd rather see this be a general interface for init modules, rather than one specific to wallet initialization.

One way to do it is to name methods to correspond to their place in the init lifecycle, like `BeforeDBOpen`, `AfterDBClose`, `BeforeNetStart`, `AfterNetStop`, etc.

jnewbery commented:
> I'd rather see this be a general interface for init modules

Do you think that's a hard requirement before this PR gets merged? This PR is simply changing function calls to an internal interface class, which can be updated or replaced later. It seems to me that creating a brand new general interface for modules is outside the scope of this.

Just merging this PR has a real benefit now - it removes one of the last remaining server -> wallet dependencies.

jnewbery commented:
Pushed a new branch. The only changes are in the final commit, which addresses @jimpo's comments above.

l2a5b1 reviewed (commented):
Concept ACK.

jimpo commented:
@jnewbery Hmm, I see the benefit, but generally when introducing interfaces I'd rather design it well up front (as opposed to, say, leaving a hack in some implementation until it can be cleaned up later). Especially if there's another proposal for a similar, general interface in https://github.com/bitcoin/bitcoin/pull/10973, I think this could warrant some more thought.

@ryanofsky Thoughts on how easy/likely it will be to switch the WalletInitInterface out for a more general one at a future date?

jnewbery commented on src/walletinitinterface.h:
```diff
@@ -0,0 +1,35 @@
+class CScheduler;
+class CRPCTable;
+
+class WalletInitInterface {
```
Thanks! Done.

jnewbery commented on src/walletinitinterface.h:
```diff
@@ -0,0 +1,35 @@
+
+class WalletInitInterface {
+public:
+    /** Get wallet help string */
```
Added

jnewbery commented:
Addressed review comments from @251Labs.

251Labs commented:
Thanks @jnewbery! ACK caaf972.

jnewbery commented:
Thanks for the review @251Labs. I've added a fixup commit to add override specifiers to the member functions in `WalletInit`. Let me know what you think. I'll squash it into the previous commit if you agree.

jamesob commented:
Tested ACK https://github.com/bitcoin/bitcoin/pull/10762/commits/b3e482c65c5c4697b38bba6006dbcdc3c1bba836

- [x] Encrypt existing wallet (*Settings -> Encrypt Wallet...*)
- [x] Change wallet passphrase (*Settings -> Change Passphrase...*)
- [x] Tested sending a transaction to self with automatic coin selection (*Send*)
- [x] Backup wallet & start with `-wallet=/path/to/backup` (*File -> Backup wallet...)

ryanofsky reviewed (commented):
utACK b3e482c65c5c4697b38bba6006dbcdc3c1bba836. Changes since previous review: removing Wallet from `WalletInit` method names, switching to pure virtual methods, removing `-disablewallet` check in StopWallets, fixing GetWalletHelpString bug.

ryanofsky reviewed (commented):
utACK d8a53d0eb5d168947544d2b2ce6abccef4e5950c for new commit adding `override`s

TheBlueMatt commented on src/init.h:
```diff
@@ -6,11 +6,15 @@
 class CWallet;
 
+class WalletInitInterface;
+extern std::unique_ptr<WalletInitInterface> g_wallet_init_interface;
```
Hmm, can we avoid yet more globals and just pass the wallet_init_interface in to the AppInit* functions?

TheBlueMatt commented on src/init.cpp:
```diff
@@ -189,9 +188,9 @@ void Shutdown()
-#ifdef ENABLE_WALLET
-    FlushWallets();
-#endif
+    if (g_wallet_init_interface) {
```
Can we just make this always non-nullptr and call it directly? The mess of if (g_...) g_...->DoThing() in init.cpp is starting to get very repetitive.

ryanofsky replied:
FWIW, in my followup, there can be multiple clients, so the if statements turn into for loops. Would be fine to change this, but it's unclear what the benefits would be.

jimpo commented:
utACK d8a53d0eb5d168947544d2b2ce6abccef4e5950c

jnewbery commented:
Squashed the fixup commit. d8a53d0 -> 49baa4a462193d8d82b51d464740aa5f1114edf1 should be identical

jnewbery committed:
[wallet] Add dummy wallet init class

jnewbery commented:
Added a dummy wallet init class as requested here: https://github.com/bitcoin/bitcoin/pull/10762#discussion_r177495634

I started trying to implement the suggestion of passing the wallet_init_interface into the AppInit* functions, but it ended up being a much larger change than I was expecting, since the wallet interface is called from several different functions in init.cpp, which are called from various places in qt.

I'm going to leave it as a global, since it's limited to the init.cpp file, and this PR already has 4 ACKs. The global can be removed in a future PR if necessary.

Sjors commented:
I did some light testing in QT on macOS. I also rebased on master to see if multiwallet behaves, which it seems to.

Travis failure seems to be due to #12806.

PierreRochard commented on src/init.cpp:
```diff
@@ -1597,12 +1582,7 @@ bool AppInitMain()
-    if (!OpenWallets())
-        return false;
-#else
-    LogPrintf("No wallet support compiled in!\n");
```
Should this logging be added to the DummyWalletInit Open function?

jnewbery replied:
Good idea. This PR seems to have enough ACKs for merging. Let's save this for a follow-up PR?

MarcoFalke replied:
^ This is now up for grabs

TheBlueMatt commented:
utACK c7ec5243892c38f9f77781b0e24a237942e7c776

PierreRochard commented:
Tested ACK c7ec5243892c38f9f77781b0e24a237942e7c776

jamesob commented:
reACK https://github.com/bitcoin/bitcoin/pull/10762/commits/c7ec5243892c38f9f77781b0e24a237942e7c776

ajtowns commented:
utACK c7ec5243892c38f9f77781b0e24a237942e7c776

promag commented:
Tested ACK c7ec524.

@jnewbery please see b3bcb3f, IMO it improves the separation and hides the interface implementations `WalletInitInterface` and `DummyWalletInit`.

jnewbery commented:
@promag thanks for the review. Can we leave your suggestion for a follow-up PR in order not to invalidate the ACKs?

laanwj commented:
This is nice. utACK c7ec5243892c38f9f77781b0e24a237942e7c776
